src/AvlTree/AvlTreeNode.ts

```diff
@@ -262,9 +262,33 @@ class AvlTreeNode<ValueType> {
   }
 
   public findRange(beginValue: ValueType, endValue: ValueType): number[] | null {
-    return new Array<number>();
+    const selectedValues = new Array<number>();
+
+    return this.searchRange(selectedValues, beginValue, endValue);
   }
 
+  public searchRange(selectedValues: number[], beginValue: ValueType, endValue: ValueType): number[] | null {
+    const beginComparisonResult = this.comparator.compare(this.value, beginValue);
+    const endComparisonResult = this.comparator.compare(this.value, endValue);
+
+    if (beginComparisonResult < 0 && this.rightChildren) {
+      return this.rightChildren.searchRange(selectedValues, beginValue, endValue);
+    } else {
+      if (endComparisonResult <= 0) {
+        selectedValues.push(...this.indexes);
+        if (this.leftChildren) {
+          return this.leftChildren.searchRange(selectedValues, beginValue, endValue);
+        } else if (this.rightChildren) {
+          return this.rightChildren.searchRange(selectedValues, beginValue, endValue);
+        }
+      } else {
+        if (this.leftChildren) {
+          return this.leftChildren.searchRange(selectedValues, beginValue, endValue);
+        }
+      }
+    }
+    return selectedValues;
+  }
```

Comment:
I would suggest following a structure like below, to avoid nested if's, and (in my opinion) make the flow more clear
* If the value is between the begin and end Values: push the indexes and search in left AND right children;
* If the value greater than the end Values and have right children: search in the right children;
* If the value smaller than the begin Values and have left children: search in the left children;

This code is done in a rush, so the logic may be messy, but it's the structure I find more optimal.

```suggestion
  if (beginComparisonResult >= 0 && endComparisonResult <= 0) {
    selectedValues.push(...this.indexes);
    selectedValues.push(...this.leftChildren.searchRange(selectedValues, beginValue, endValue));
    selectedValues.push(...this.rightChildren.searchRange(selectedValues, beginValue, endValue));
  }else if (endComparisonResult > 0 && this.rightChildren) {
    return this.rightChildren.searchRange(selectedValues, beginValue, endValue);
  } else if (beginComparisonResult < 0 && this.leftChildren) {
    return this.leftChildren.searchRange(selectedValues, beginValue, endValue);
  }
  return selectedValues;
```

Reply:
I have updated the branch with the suggestions, please check the latest update.